Sources/Basic/OutputByteStream.swift

```diff
@@ -43,7 +43,93 @@ public protocol ByteStreamable {
+    func write(_ byte: UInt8)
+
+    /// Write a collection of bytes to the buffer.
+    func write<C: Collection>(_ bytes: C) where C.Element == UInt8
+
+    /// Flush the stream's buffer.
+    func flush()
+}
+
+extension OutputByteStream {
+    /// Write a sequence of bytes to the buffer.
+    public func write<S: Sequence>(sequence: S) where S.Iterator.Element == UInt8 {
+        // Iterate the sequence and append byte by byte since sequence's append
+        // is not performant anyway.
+        for byte in sequence {
+            write(byte)
+        }
+    }
+
+    /// Write a string to the buffer (as UTF8).
+    public func write(_ string: String) {
+        // FIXME(performance): Use `string.utf8._copyContents(initializing:)`.
+        write(string.utf8)
+    }
+
+    /// Write a string (as UTF8) to the buffer, with escaping appropriate for
+    /// embedding within a JSON document.
+    ///
+    /// NOTE: This writes the literal data applying JSON string escaping, but
+    /// does not write any other characters (like the quotes that would surround
+    /// a JSON string).
+    public func writeJSONEscaped(_ string: String) {
+        // See RFC7159 for reference: https://tools.ietf.org/html/rfc7159
+        for character in string.utf8 {
+            // Handle string escapes; we use constants here to directly match the RFC.
+            switch character {
+            // Literal characters.
+            case 0x20...0x21, 0x23...0x5B, 0x5D...0xFF:
+                write(character)
+
+            // Single-character escaped characters.
+            case 0x22: // '"'
+                write(0x5C) // '\'
+                write(0x22) // '"'
+            case 0x5C: // '\\'
+                write(0x5C) // '\'
+                write(0x5C) // '\'
+            case 0x08: // '\b'
+                write(0x5C) // '\'
+                write(0x62) // 'b'
+            case 0x0C: // '\f'
+                write(0x5C) // '\'
+                write(0x66) // 'b'
+            case 0x0A: // '\n'
+                write(0x5C) // '\'
+                write(0x6E) // 'n'
+            case 0x0D: // '\r'
+                write(0x5C) // '\'
+                write(0x72) // 'r'
+            case 0x09: // '\t'
+                write(0x5C) // '\'
+                write(0x74) // 't'
+
+            // Multi-character escaped characters.
+            default:
+                write(0x5C) // '\'
+                write(0x75) // 'u'
+                write(hexdigit(0))
+                write(hexdigit(0))
+                write(hexdigit(character >> 4))
+                write(hexdigit(character & 0xF))
+            }
+        }
+    }
+}
+
+/// The `OutputByteStream` base class.
+///
+/// This class provides a base and efficient implementation of the `OutputByteStream`
+/// protocol. It can not be used as is-as subclasses as several functions need to be
+/// implemented in subclasses.
+public class OutputByteStreamBase: OutputByteStream {
```

Comment:
I wonder if we should call this `_OutputByteStreamBase`.

Reply:
Why would you prefix it with an underscore? Do we really want to discourage users from creating subclasses?

Reply:
To indicate the clients that this is an implementation detail and they shouldn't use this type directly, otherwise someone might do something like:

let foo: OutputByteStreamBase = someStream

Reply:
Is there something ultimately wrong with declaring a variable of type `OutputByteStreamBase`? I don't feel strongly either way, but the Standard Library usually prefixes types with underscore only when they ought to be `private` but a missing language feature forces them to make it `public`. Is it also the case here? Or do we want to encourage people who make a custom `OutputByteStream` to subclass `OutputByteStreamBase` to profit from it's natural ability to be queued on a single dispatch queue?

Reply:
I think of `OutputByteStreamBase` as an implementation detail to share code between the multiple types of streams we have. However, we can't make this class private or internal because its subclasses are public so its kind of a like a missing language feature (although we'll probably never see such a feature). We can use composition to avoid exposing this type but I don't know if that'll be easy/worth to do. Currently, other modules can't subclass this type anyway because its not `open`. In future, if we do want to allow clients in other modules to re-use this class, we can make that change.

Reply:
You've convinced me :-) I'll prefix it.